### PR TITLE
feat(router-plugin): add esbuild plugin

### PR DESCRIPTION
**Pull request by benmcginnis**
Addresses #2576 by adding an esbuild plugin to the router-plugin package.

Duplicated the basic-file-based e2e test to verify validate the behavior, and replaces the vite config with an esbuild config as well as replacing the vite scripts with esbuild scripts.

tested by running: pnpm run test:e2e from within the e2e/react-router/basic-esbuild-file-based directory.

### CODE REVIEW

**Comment by schiller-manuel** [comment]
please duplicate the e2e project folder (name it `basic-esbuild-file-based`)

File: e2e/react-router/basic-esbuild-file-based/src/esbuild.config.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import {TanStackRouterEsbuild} from '@tanstack/router-plugin/esbuild';
+
+export default {
+  jsx: 'transform',
+  minify: true,
+  sourcemap: true,
+  bundle: true,
+  format: 'esm',
+  target: ['esnext'],
+  plugins: [TanStackRouterEsbuild()]
```
Comment (schiller-manuel) [inline e2e/react-router/basic-esbuild-file-based/src/esbuild.config.js:10]: please enable automatic code splitting

**Review by SeanCassiere — CHANGES_REQUESTED** [review]
Two things of note:

1. In the `examples/react` folder, please create a `quickstart-esbuild-file-based` example. The only difference should be the bundler, the actual source routes should remain the same as rspack, webpack, and vite examples.
2. Update the documentation in `docs/framework/react/guide/file-based-routing.md`, to:
  - Add esbuild into the list of supported bundlers.
  - Add the configuration steps for esbuild, like done for the other bundlers.

**Review by SeanCassiere — APPROVED** [review]
The changes look good.

Now just need to figure out why CI is failing.

**Pull request merged**